client/asset/eth/eth_test.go

```diff
@@ -883,6 +824,396 @@ func TestPreSwap(t *testing.T) {
+
+		expectedFees := gasNeededForSwaps(len(swaps.Contracts)) * swaps.FeeRate
+		if feeSpent != expectedFees {
+			t.Fatalf("%v: expected fees: %v != actual fees %v", testName, expectedFees, feeSpent)
+		}
+	}
+
+	secret := encode.RandomBytes(32)
+	secretHash := sha256.Sum256(secret)
+	secret2 := encode.RandomBytes(32)
+	secretHash2 := sha256.Sum256(secret2)
+	expiration := time.Now().Add(time.Hour * 8).Unix()
+
+	// Ensure error with invalid secret hash
+	contracts := []*asset.Contract{
+		{
+			Address:    receivingAddress,
+			Value:      ethToGwei(1),
+			SecretHash: encode.RandomBytes(31),
+			LockTime:   uint64(expiration),
+		},
+	}
+	inputs := refreshWalletAndFundCoins(5, []uint64{ethToGwei(2)})
+	swaps := asset.Swaps{
+		Inputs:     inputs,
+		Contracts:  contracts,
+		FeeRate:    200,
+		LockChange: false,
+	}
+	testSwap("incorrect length secret hash", swaps, true)
+
+	// Ensure error with invalid receiving address
+	contracts = []*asset.Contract{
+		{
+			Address:    hex.EncodeToString(encode.RandomBytes(21)),
+			Value:      ethToGwei(1),
+			SecretHash: secretHash[:],
+			LockTime:   uint64(expiration),
+		},
+	}
+	inputs = refreshWalletAndFundCoins(5, []uint64{ethToGwei(2)})
+	swaps = asset.Swaps{
+		Inputs:     inputs,
+		Contracts:  contracts,
+		FeeRate:    200,
+		LockChange: false,
+	}
+	testSwap("invalid receiving address", swaps, true)
+
+	// Ensure error when initializing swap errors
+	node.initErr = errors.New("")
+	contracts = []*asset.Contract{
+		{
+			Address:    receivingAddress,
+			Value:      ethToGwei(1),
+			SecretHash: secretHash[:],
+			LockTime:   uint64(expiration),
+		},
+	}
+	inputs = refreshWalletAndFundCoins(5, []uint64{ethToGwei(2)})
+	swaps = asset.Swaps{
+		Inputs:     inputs,
+		Contracts:  contracts,
+		FeeRate:    200,
+		LockChange: false,
+	}
+	testSwap("error initialize but no send", swaps, true)
+	node.initErr = nil
+
+	// Ensure error initializing two contracts with same secret hash
+	contracts = []*asset.Contract{
+		{
+			Address:    receivingAddress,
+			Value:      ethToGwei(1),
+			SecretHash: secretHash[:],
+			LockTime:   uint64(expiration),
+		},
+		{
+			Address:    receivingAddress,
+			Value:      ethToGwei(1),
+			SecretHash: secretHash[:],
+			LockTime:   uint64(expiration),
+		},
+	}
+	inputs = refreshWalletAndFundCoins(5, []uint64{ethToGwei(3)})
+	swaps = asset.Swaps{
+		Inputs:     inputs,
+		Contracts:  contracts,
+		FeeRate:    200,
+		LockChange: false,
+	}
+	testSwap("two contracts same hash error", swaps, true)
+
+	// Tests one contract without locking change
+	contracts = []*asset.Contract{
+		{
+			Address:    receivingAddress,
+			Value:      ethToGwei(1),
+			SecretHash: secretHash[:],
+			LockTime:   uint64(expiration),
+		},
+	}
+	inputs = refreshWalletAndFundCoins(5, []uint64{ethToGwei(2)})
+	swaps = asset.Swaps{
+		Inputs:     inputs,
+		Contracts:  contracts,
+		FeeRate:    200,
+		LockChange: false,
+	}
+	testSwap("one contract, don't lock change", swaps, false)
```

Comment:
Probably OK, but we really shouldn't need a change coin if `LockChange` is `false`. **core** shouldn't require it. It does look like we are returning the change coin for other assets. I don't think that's on purpose.